controllers/capability/reconciler.go

```diff
@@ -0,0 +1,249 @@
+package capability
+
+import (
+	"context"
+	"hash/fnv"
+	"strconv"
+
+	"github.com/Dynatrace/dynatrace-operator/api/v1alpha1"
+	"github.com/Dynatrace/dynatrace-operator/controllers/customproperties"
+	"github.com/Dynatrace/dynatrace-operator/controllers/dtpullsecret"
+	"github.com/Dynatrace/dynatrace-operator/controllers/dtversion"
+	"github.com/Dynatrace/dynatrace-operator/controllers/kubesystem"
+	"github.com/Dynatrace/dynatrace-operator/controllers/utils"
+	"github.com/Dynatrace/dynatrace-operator/dtclient"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type Reconciler struct {
+	client.Client
+	Instance                         *v1alpha1.DynaKube
+	apiReader                        client.Reader
+	scheme                           *runtime.Scheme
+	dtc                              dtclient.Client
+	log                              logr.Logger
+	imageVersionProvider             dtversion.ImageVersionProvider
+	enableUpdates                    bool
+	module                           string
+	capabilityName                   string
+	serviceAccountOwner              string
+	capability                       *v1alpha1.CapabilityProperties
+	onAfterStatefulSetCreateListener []StatefulSetEvent
+}
+
+func NewReconciler(clt client.Client, apiReader client.Reader, scheme *runtime.Scheme, dtc dtclient.Client, log logr.Logger,
+	instance *v1alpha1.DynaKube, imageVersionProvider dtversion.ImageVersionProvider, enableUpdates bool,
+	capability *v1alpha1.CapabilityProperties, module string, capabilityName string, serviceAccountOwner string) *Reconciler {
+	if serviceAccountOwner == "" {
+		serviceAccountOwner = module
+	}
+
+	return &Reconciler{
+		Client:                           clt,
+		apiReader:                        apiReader,
+		scheme:                           scheme,
+		dtc:                              dtc,
+		log:                              log,
+		Instance:                         instance,
+		imageVersionProvider:             imageVersionProvider,
+		enableUpdates:                    enableUpdates,
+		module:                           module,
+		capabilityName:                   capabilityName,
+		serviceAccountOwner:              serviceAccountOwner,
+		capability:                       capability,
+		onAfterStatefulSetCreateListener: []StatefulSetEvent{},
+	}
+}
+
+func (r *Reconciler) AddOnAfterStatefulSetCreateListener(event StatefulSetEvent) {
+	r.onAfterStatefulSetCreateListener = append(r.onAfterStatefulSetCreateListener, event)
```

Comment:
I agree that there are a lot of abstraction going on, but this way has the advantage, that specific values can be modified without much work.
If we added a function in a specific reconciler, we would have to check if the sts exists, get the sts, change it, recalculate the hash and finally update it. 
Using this event removes all these steps to only having to change specific values, since it fires before the hash is calculated.
So although it is more abstract, I think it is cleaner.

Reply:
I was referring more to have a single parameter with a callback function. I agree with the point of needing custom changes based on the feature, but the array is a perhaps a bit too much, and just a single callback does the job.

I'll leave it to you.